### PR TITLE
DBZ-2365 Support for crss-db datatypes

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.redhat.com/browse/DBZ-2365

### CODE REVIEW

**Comment by jpechane** [comment]
Upstream PR - https://github.com/antlr/grammars-v4/pull/1794

File: debezium-ddl-parser/src/main/antlr4/io/debezium/ddl/parser/mysql/generated/MySqlParser.g4 [hunk 1]
```diff
@@ -2045,10 +2045,10 @@ constant
 //    Data Types
 
 dataType
-    : typeName=(
+    : (typeName=(
       CHAR | CHARACTER | VARCHAR | TINYTEXT | TEXT | MEDIUMTEXT | LONGTEXT
-       | NCHAR | NVARCHAR
-      )
+       | NCHAR | NVARCHAR | LONG
+      ) | typeName=LONG VARCHAR)                                    // LONG VARCHAR is the same as LONG
```
Comment (rk3rn3r) [inline debezium-ddl-parser/src/main/antlr4/io/debezium/ddl/parser/mysql/generated/MySqlParser.g4:2051]: LONG VARCHAR equals MEDIUMTEXT

File: debezium-ddl-parser/src/main/antlr4/io/debezium/ddl/parser/mysql/generated/MySqlParser.g4 [hunk 2]
```diff
@@ -2060,22 +2060,23 @@ dataType
...
     | typeName=(
         DATE | TINYBLOB | BLOB | MEDIUMBLOB | LONGBLOB
         | BOOL | BOOLEAN | SERIAL
       )                                                             #simpleDataType
-    | typeName=(
+    | (typeName=(
         BIT | TIME | TIMESTAMP | DATETIME | BINARY
         | VARBINARY | YEAR
-      )
+      ) | LONG typeName=VARBINARY)                                  // LONG VARBINARY is the same as VARBINARY
```
Comment (rk3rn3r) [inline debezium-ddl-parser/src/main/antlr4/io/debezium/ddl/parser/mysql/generated/MySqlParser.g4:2079]: - LONG VARBINARY equals MEDIUMBLOB
- same here for the lexer name?

File: debezium-ddl-parser/src/main/antlr4/io/debezium/ddl/parser/mysql/generated/MySqlLexer.g4 [hunk 3]
```diff
@@ -248,6 +256,7 @@ VARBINARY:                           'VARBINARY';
 TINYBLOB:                            'TINYBLOB';
 BLOB:                                'BLOB';
 MEDIUMBLOB:                          'MEDIUMBLOB';
+LONG:                                'LONG';
```
Comment (rk3rn3r) [inline debezium-ddl-parser/src/main/antlr4/io/debezium/ddl/parser/mysql/generated/MySqlLexer.g4:259]: - shouldn't the lexer name be closer to LONG VARCHAR/VARBINARY ? maybe separate tokens?

File: debezium-ddl-parser/src/main/antlr4/io/debezium/ddl/parser/mysql/generated/MySqlParser.g4 (hunk 1 above)
Comment (jpechane) [reply debezium-ddl-parser/src/main/antlr4/io/debezium/ddl/parser/mysql/generated/MySqlParser.g4:2051]: See https://dev.mysql.com/doc/refman/8.0/en/other-vendor-data-types.html
`LONG` == `LONG VARCHAR` and this maps to `MEDIUMTEXT` - so I opted for easier grammar as as soon as you'll get `LONG` it does not matter if there is `VARCHAR` or not

File: debezium-ddl-parser/src/main/antlr4/io/debezium/ddl/parser/mysql/generated/MySqlParser.g4 (hunk 2 above)
Comment (jpechane) [reply debezium-ddl-parser/src/main/antlr4/io/debezium/ddl/parser/mysql/generated/MySqlParser.g4:2079]: You ar eprobably right here - I implemented this in narrow Debezium view as from Debezium POV VARIBIANRY is the same as MEDIUMBLOB. But from general/upstream POV it should differ.

Comment (rk3rn3r) [reply debezium-ddl-parser/src/main/antlr4/io/debezium/ddl/parser/mysql/generated/MySqlParser.g4:2079]: agree. VARCHAR might be related to a encoding, like unicode, where BLOB might be some arbitrary raw bytes. DBZ doesn't care yet, correct?

**Comment by jpechane** [comment]
@rk3rn3r @gunnarmorling This is ready for new review - parser has the separate rules for `LONG...`

**Comment by rk3rn3r** [comment]
Looks good to me now. but I need to confess all this types stuff is hard to oversee. :crossed_fingers:
maybe @gunnarmorling can have a look too and then merge?

**Comment by gunnarmorling** [comment]
LGTM, merging. Thanks for fixing/reviewing, @jpechane, @rk3rn3r. Interesting that this hasn't been reported before, seems not to be used that widely after all.

**Pull request merged**